src/borg/archive.py

```diff
@@ -1117,10 +1117,16 @@ def defer_if_necessary(item1, item2):
             update_hardlink_masters(deleted, deleted_item)
             yield (path, compare_items(deleted, deleted_item))
         for item1, item2 in deferred:
-            assert hardlink_master_seen(item1)
-            assert hardlink_master_seen(item2)
             assert item1.path == item2.path, "Deferred items have different paths"
-            yield (item1.path, compare_items(item1, item2))
+            hl_notfound1 = not hardlink_master_seen(item1)
+            hl_notfound2 = not hardlink_master_seen(item2)
+            if hl_notfound1 or hl_notfound2:
+                if hl_notfound1:
+                    print_warning(f"cannot find hardlink source for {item1.path} ({item1.source}), skipping compare.")
+                if hl_notfound2:
+                    print_warning(f"cannot find hardlink source for {item2.path} ({item2.source}), skipping compare.")
+            else:
+                yield (item1.path, compare_items(item1, item2))
```

Comment:
can we rearrange this a bit, like:
- use hl_found(1|2) as variables
- first check for the desired case "hl_found1 and hl_found2" and yield
- then have an else branch and in there, check for not hl_found1 for the hl1 warning and for not hl_found2 for the hl2 warning.